server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java

```diff
@@ -49,7 +58,64 @@ public static FetchSourceContext of(boolean fetchSource, @Nullable String[] incl
         if ((includes == null || includes.length == 0) && (excludes == null || excludes.length == 0)) {
             return of(fetchSource);
         }
-        return new FetchSourceContext(fetchSource, includes, excludes);
+        return new FetchSourceContext(fetchSource, includes, excludes, DEFAULT_INCLUDE_VECTORS);
+    }
+
+    public static FetchSourceContext of(boolean fetchSource, @Nullable String[] includes, @Nullable String[] excludes, @Nullable Boolean includeVectors) {
+        if ((includes == null || includes.length == 0) && (excludes == null || excludes.length == 0) && includeVectors == DEFAULT_INCLUDE_VECTORS) {
+            return of(fetchSource);
+        }
+        return new FetchSourceContext(fetchSource, includes, excludes, includeVectors);
+    }
+
+    public static FetchSourceContext of(FetchSourceContext original, IndexShard indexShard) {
+        MappingLookup mappingLookup = indexShard.mapperService().mappingLookup();
+        if (original.includeVectors() == null) {
+            String[] inferenceFields = mappingLookup
+                .inferenceFields()
+                .keySet()
+                .toArray(String[]::new);
+
+            return FetchSourceContext.of(
+                original.fetchSource(),
+                original.includes(),
+                ArrayUtils.concat(original.excludes(), inferenceFields)
+            );
+        } else if (original.includeVectors()) {
+            return original;
+        } else {
+            String[] excludeFields = original.excludes();
+            String[] inferenceFields = mappingLookup
+                .inferenceFields()
+                .keySet()
+                .toArray(String[]::new);
+
+            excludeFields = ArrayUtils.concat(excludeFields, inferenceFields);
+
+            String[] denseVectors = mappingLookup
+                .getFullNameToFieldType()
+                .entrySet()
+                .stream()
+                .filter(entry -> entry.getValue() instanceof DenseVectorFieldMapper.DenseVectorFieldType)
+                .map(Map.Entry::getKey)
+                .toArray(String[]::new);
+            excludeFields = ArrayUtils.concat(excludeFields, denseVectors);
+
+            String[] sparseVectors = mappingLookup
+                .getFullNameToFieldType()
+                .entrySet()
+                .stream()
+                .filter(entry -> entry.getValue() instanceof SparseVectorFieldMapper.SparseVectorFieldType)
+                .map(Map.Entry::getKey)
+                .toArray(String[]::new);
+            excludeFields = ArrayUtils.concat(excludeFields, sparseVectors);
+
+            return FetchSourceContext.of(
+                original.fetchSource(),
+                original.includes(),
+                excludeFields
+            );
+        }
```

Comment:
As we are trying to mutate before `filter` is called, I could not find a better way to apply my changes here. https://github.com/elastic/elasticsearch/blob/6ca3ac253a8d8171f228d9dbfd6c5c924239c226/server/src/main/java/org/elasticsearch/action/update/UpdateHelper.java#L346

`UpdateHelper` and `UpdateRequest` do not have access to `indexShard`. Any suggestion if and how I need to handle this scenario?